### PR TITLE
[gcp-auth-extension]: Try resolving GCP_PROJECT with ServiceOptions if not provided

### PR DESCRIPTION
**Pull request by gustavovnicius**
**Description:**

Enhancement - Try resolving GOOGLE_CLOUD_PROJECT value using google-cloud-sdk.

Having a mandatory `GOOCLE_CLOUD_PROJECT` configuration is unnecessary, given the extension already depends on ADC context existing. It can try to infer the project from it.
Still mandatory to have a project id, but this gives more flexibility in the usage of the extension.

**Existing Issue(s):**
https://github.com/open-telemetry/opentelemetry-java-contrib/issues/2102

**Testing:**

Refactored the project id resolver logic into a separate method, including the more comprehensive `ServiceOptions` resolver. Tested the method for both the current strategy and falling back to `ServiceOptions`, also added a missing test for not being able to find the option and throwing an exception.

**Documentation:**

`GOOGLE_CLOUD_PROJECT` configuration was moved into the optional config section, with a note that even though it's optional, the extension needs it either provided or being able to infer it. 

**Outstanding items:**

N/A

### CODE REVIEW

**Comment by gustavovnicius** [comment]
Any outstanding blocker about this one?

**Comment by trask** [comment]
cc @psx95 @jsuereth

File: gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java [hunk 1]
```diff
@@ -538,6 +541,49 @@ void testTargetSignalsBehavior(TargetSignalBehavior testCase) {
...
+        GcpAuthAutoConfigurationCustomizerProvider.getGoogleProjectId(configProperties));
+  }
+
+  @Test
+  void testResolveGoogleProjectIdFromServiceOptions() {
+    // Clear the system property to ensure it does not interfere with the test
+    System.clearProperty(ConfigurableOption.GOOGLE_CLOUD_PROJECT.getSystemProperty());
+    // Property that the extension expects is google.cloud.project, ServiceOptions works with
+    // GOOGLE_CLOUD_PROJECT, so this will highlight the fallback to ServiceOptions
+    System.setProperty("GOOGLE_CLOUD_PROJECT", DUMMY_GCP_RESOURCE_PROJECT_ID);
```
Comment (laurit) [inline gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java]: consider using junit `SetSystemProperty` and `ClearSystemProperty` annotations. Then you won't need to worry about properties set in one test messing with the next test.

Comment (gustavovnicius) [reply gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java]: Fixed

**Comment by gustavovnicius** [comment]
New version has been adopted and project id is attempted to be resolved based on credentials. @psx95

File: gcp-auth-extension/README.md [hunk 2]
```diff
@@ -34,14 +34,13 @@ The extension can be configured either by environment variables or system proper
...
-#### Required Config
+#### Optional Config
 
 - `GOOGLE_CLOUD_PROJECT`: Environment variable that represents the Google Cloud Project ID to which the telemetry needs to be exported.
 
   - Can also be configured using `google.cloud.project` system property.
-  - This is a required option, the agent configuration will fail if this option is not set.
-
-#### Optional Config
+  - If neither of these options are set, the extension will attempt to infer the project id from the credentials used, however notice that not all credentials implementations will be able to provide a project id, so the inference is only a best-effort attempt.
```
Comment (psx95) [inline gcp-auth-extension/README.md]:
```suggestion
  - If neither of these options are set, the extension will attempt to infer the project id from the current credentials as a fallback, however notice that not all credentials implementations will be able to provide a project id, so the inference is only a best-effort attempt.
```

File: gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java [hunk 3]
```diff
@@ -228,10 +234,36 @@ private static Map<String, String> getRequiredHeaderMap(
...
+
+  /**
+   * Retrieves the Google Cloud Project ID from the configuration properties, falling back to
+   * google-cloud-core's ServiceOptions project ID resolution if not explicitly set.
+   *
+   * @param configProperties The configuration properties containing the GCP project ID.
+   * @return The Google Cloud Project ID.
+   */
+  @Nonnull
+  static String getGoogleProjectId(
```
Comment (psx95) [inline gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java]: We could do away with this functions since much of this functionality is already available in `ConfigurableOption` class.

See suggestion for `customizeResource` method.

File: gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java [hunk 4]
```diff
@@ -228,10 +234,36 @@ private static Map<String, String> getRequiredHeaderMap(
...
-  private static Resource customizeResource(Resource resource, ConfigProperties configProperties) {
-    String gcpProjectId =
-        ConfigurableOption.GOOGLE_CLOUD_PROJECT.getConfiguredValue(configProperties);
-    Resource res = Resource.create(Attributes.of(stringKey(GCP_USER_PROJECT_ID_KEY), gcpProjectId));
+  private static Resource customizeResource(Resource resource, String gcpProjectId) {
+    Resource res =
+        Resource.create(
+            Attributes.of(AttributeKey.stringKey(GCP_USER_PROJECT_ID_KEY), gcpProjectId));
     return resource.merge(res);
   }
```
Comment (psx95) [inline gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java]:
```suggestion
  // Note that credentials can be passed from `customize` function directly
  private static Resource customizeResource(Resource resource, ConfigProperties configProperties, GoogleCredentials credentials) {
    String gcpProjectId;
    try {
      gcpProjectId = ConfigurableOption.GOOGLE_CLOUD_PROJECT.getConfiguredValue(configProperties);
    } catch (GoogleAuthException e) {
      gcpProjectId = credentials.getProjectId();
      if (gcpProjectId == null) {
        // this exception will still contain the accurate message.
        throw e;  
      }      
    }
    Resource res = Resource.create(Attributes.of(stringKey(GCP_USER_PROJECT_ID_KEY), gcpProjectId));
    return resource.merge(res);
  }
```

File: gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java [hunk 5]
```diff
@@ -538,6 +543,51 @@ void testTargetSignalsBehavior(TargetSignalBehavior testCase) {
...
+    Mockito.verifyNoInteractions(mockedGoogleCredentials);
+  }
+
+  @Test
+  @ClearSystemProperty(key = "google.cloud.project")
+  @ClearSystemProperty(key = "GOOGLE_CLOUD_PROJECT")
+  void testResolveGoogleProjectIdFromCredentials() {
+    DefaultConfigProperties configProperties =
+        DefaultConfigProperties.create(Collections.emptyMap(), Mockito.mock(ComponentLoader.class));
+
```
Comment (psx95) [inline gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java]: nit: extra line

**Review by psx95 — COMMENTED** [review]
Thanks for the PR! 
Left a couple of suggestions that might be helpful.

File: gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java (hunk 3 above)
Comment (gustavovnicius) [reply gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java]: @psx95, this method was added mostly so we could isolate the project id resolution in the tests.
Any suggestions for that?

Comment (psx95) [reply gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java]: Hi @gustavovnicius, this method is an implementation detail of the class and should have been private. For testing, I think you should test the expected behavior instead of testing the method in isolation. I think `testQuotaProjectBehavior` does something similar. 

 So an idea here would be to mock the `GoogleCredentials` (particularly the `getProjectId` method) and in one of the cases, you return a valid Project ID and see if the extension loads or throws an expected `GoogleAuthException`

Let me know if this makes sense.

Comment (gustavovnicius) [reply gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java]: Hey @psx95 , suggestion applied. Tried to replicate the same test structure as `testQuotaProjectBehavior`

**Comment by gustavovnicius** [comment]
@psx95 ?

File: gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java [hunk 6]
```diff
@@ -228,9 +230,19 @@ private static Map<String, String> getRequiredHeaderMap(
...
 
   // Updates the current resource with the attributes required for ingesting OTLP data on GCP.
-  private static Resource customizeResource(Resource resource, ConfigProperties configProperties) {
-    String gcpProjectId =
-        ConfigurableOption.GOOGLE_CLOUD_PROJECT.getConfiguredValue(configProperties);
+  // Note that credentials can be passed from `customize` function directly
```
Comment (psx95) [inline gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java]: Nit: Can remove this comment. 

I added this in the suggestion only to clarify my reasoning.

File: gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java [hunk 7]
```diff
@@ -228,9 +230,19 @@ private static Map<String, String> getRequiredHeaderMap(
...
+  private static Resource customizeResource(
+      Resource resource, GoogleCredentials credentials, ConfigProperties configProperties) {
+    String gcpProjectId;
+    try {
+      gcpProjectId = ConfigurableOption.GOOGLE_CLOUD_PROJECT.getConfiguredValue(configProperties);
+    } catch (io.opentelemetry.sdk.autoconfigure.spi.ConfigurationException e) {
```
Comment (psx95) [inline gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java]: Nit: Can replace it with an import statement.

File: gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java [hunk 8]
```diff
@@ -228,9 +230,19 @@ private static Map<String, String> getRequiredHeaderMap(
...
+      gcpProjectId = credentials.getProjectId();
+      if (gcpProjectId == null) {
+        // this exception will still contain the accurate message.
```
Comment (psx95) [inline gcp-auth-extension/src/main/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProvider.java]: Nit: Can remove this comment. 

I added this in the suggestion only to clarify my reasoning.

File: gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java [hunk 9]
```diff
@@ -680,6 +767,34 @@ private static Stream<Arguments> provideTargetSignalBehaviorTestCases() {
    * indicates that the mocked credentials are configured to provide DUMMY_GCP_QUOTA_PROJECT_ID as
    * the quota project ID.
    */
+  private static Stream<Arguments> provideProjectIdBehaviorTestCases() {
```
Comment (psx95) [inline gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java:762]: The javadoc string from line 755 to 769 applies only to `provideQuotaBehaviorTestCases`, please move it so that it sits right above that function. 

`provideProjectIdBehaviorTestCases` would require a *similar*, but *different* doc string.

File: gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java [hunk 10]
```diff
@@ -680,6 +767,34 @@ private static Stream<Arguments> provideTargetSignalBehaviorTestCases() {
...
+        // If user doesn't specify project ID, fallback to credentials.getProjectId()
+        Arguments.of(
+            ProjectIdTestBehavior.builder()
+                .setUserSpecifiedProjectId(null)
+                .setCredentialsProjectId("credentials-project-id")
+                .setExpectedProjectIdInResource("credentials-project-id")
+                .setExpectedToThrow(false)
+                .build()),
+        // If user doesn't specify and credentials.getProjectId() returns null, throw exception
+        Arguments.of(
```
Comment (psx95) [inline gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java:789]: Some edge cases should also be tested here - particularly around the extension's behavior on empty string values (`""`) for `setUserSpecifiedProjectId` and `setCredentialsProjectId`.

File: gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java [hunk 11]
```diff
@@ -456,6 +457,92 @@ void testQuotaProjectBehavior(QuotaProjectIdTestBehavior testCase) throws IOExce
...
+  @MethodSource("provideProjectIdBehaviorTestCases")
+  @ClearSystemProperty(key = "google.cloud.project")
+  @ClearSystemProperty(key = "google.otel.auth.target.signals")
+  @SuppressWarnings("CannotMockMethod")
+  void testProjectIdBehavior(ProjectIdTestBehavior testCase) throws IOException {
+
+    // configure environment according to test case
+    String userSpecifiedProjectId = testCase.getUserSpecifiedProjectId();
+    if (userSpecifiedProjectId != null && !userSpecifiedProjectId.isEmpty()) {
+      System.setProperty(
```
Comment (psx95) [inline gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java:470]: The user can pass empty strings as the projectID, as such the test should account for it.

File: gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java [hunk 12]
```diff
@@ -456,6 +457,92 @@ void testQuotaProjectBehavior(QuotaProjectIdTestBehavior testCase) throws IOExce
...
+        Mockito.mockStatic(GoogleCredentials.class)) {
+      googleCredentialsMockedStatic
+          .when(GoogleCredentials::getApplicationDefault)
+          .thenReturn(mockedGoogleCredentials);
+
+      if (testCase.getExpectedToThrow()) {
+        // expect exception to be thrown when project ID is not available
+        assertThatThrownBy(
+                () -> {
+                  OpenTelemetrySdk sdk =
```
Comment (psx95) [inline gcp-auth-extension/src/test/java/io/opentelemetry/contrib/gcp/auth/GcpAuthAutoConfigurationCustomizerProviderTest.java]: If it is expected to throw an exception, on account of improper configuration (User project ID cannot be determined), then it should be thrown only from `buildOpenTelemetrySdkWithExporter`, right ? 

Then there is no need to invoke `generateTestSpan` or `shutdown` on the SDK - we should remove these, because this test will pass even if the exception is coming from `generateTestSpan` or `shutdown`.

**Review by psx95 — CHANGES_REQUESTED** [review]
Hi @gustavovnicius, 
Apologies for the delay on this, and thank you for the changes, they look good to me now, just suggested a few changes in the test, mostly about adding a few edge cases and updating the documentation there.

We should be good to merge after this 👍🏻 \
Thanks again for your contribution and patience!
